### PR TITLE
Support decorated providers (again)

### LINKED ISSUE #127: Support decorated providers
Currently, providers with decorators are broken:
```python
import sciline as sl
import functools

def deco(f):
    @functools.wraps(f)
    def impl(*args, **kwargs):
        return f(*args, **kwargs)
    return impl

@deco
def foo(x: int) -> str:
    return str(x)

pl = sl.Pipeline([foo], params={int: 3})
pl.compute(str)
```
raises 
```
TypeError: foo() missing 1 required positional argument: 'x'
```

I think this was introduced in #116 because
```python
import functools
import inspect
from typing import get_type_hints

def deco(f):
    @functools.wraps(f)
    def impl(*args, **kwargs):
        return f(*args, **kwargs)
    return impl

@deco
def foo(x: int) -> str:
    return str(x)

print(get_type_hints(foo))
# -> {'x': <class 'int'>, 'return': <class 'str'>}
print(inspect.getfullargspec(foo))
# -> FullArgSpec(args=[], varargs='args', varkw='kwargs', defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={'return': <class 'str'>})
```
So using `getfullargspec` is not enough, we also need to use `get_type_hints`.

Unfortunately, this breaks keyword-only args:
```python
@deco
def foo(x: int, *k: float) -> str:
    return str(x)

print(get_type_hints(foo))
# -> {'x': <class 'int'>, 'k': <class 'float'>, 'return': <class 'str'>}
print(inspect.getfullargspec(foo))
# -> FullArgSpec(args=[], varargs='args', varkw='kwargs', defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={'return': <class 'str'>})
```

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #127

### CODE REVIEW

**Comment by YooSunYoung** [comment]
I added more tests and opened a PR into this branch: #132

**Comment by jl-wynen** [comment]
#132 does not look like it would block this PR because it doesn't test anything new.

**Comment by YooSunYoung** [comment]
No it won't... just wanted to target to this branch cuz it needed this update.

**Review by YooSunYoung — APPROVED** [review]

File: docs/recipes/applying-decorators.ipynb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    "\n",
+    "For Python 3.10+, the decorator itself can be type-annoted like this:\n",
+    "\n",
+    "```python\n",
+    "from typing import ParamSpec, TypeVar\n",
+    "\n",
+    "P = ParamSpec('P')\n",
+    "R = TypeVar('R')\n",
+    "\n",
+    "def deco(f: Callable[P R]) -> Callable[P, R]:\n",
```
Comment (SimonHeybrock) [inline docs/recipes/applying-decorators.ipynb:100]: Is there a comma missing here?

**Review by SimonHeybrock — COMMENTED** [review]
I suppose we should make a bugfix release for this?

File: docs/recipes/applying-decorators.ipynb (hunk 1 above)
Comment (jl-wynen) [reply docs/recipes/applying-decorators.ipynb:100]: See #134

**Comment by jl-wynen** [comment]
> I suppose we should make a bugfix release for this?

Yes

**Pull request merged**